### PR TITLE
algebra: authenticated-poly: Implement shared polynomial floor division

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR implements polynomial floor division between two polynomials that are shared. This is done in a constant round implementation given by [Mohassel and Franklin](https://iacr.org/archive/pkc2006/39580045/39580045.pdf).

To support this, a few other primitives are added:
- Polynomial multiplicative inversion in the quotient ring $\mathbb{F}[x] / (x^n)$
- Shared scalar multiplicative inversion

### Testing
- Unit and integration tests pass
- Added unit tests for all new primitives

### CODE REVIEW

File: src/algebra/poly/authenticated_poly.rs [hunk 1]
```diff
@@ -379,6 +476,65 @@ impl<C: CurveGroup> Div<&DensePolynomialResult<C>> for &AuthenticatedDensePoly<C
 }
 impl_borrow_variants!(AuthenticatedDensePoly<C>, Div, div, /, DensePolynomialResult<C>, C: CurveGroup);
 
+/// Authenticated division, i.e. division in which the divisor is a secret shared polynomial
+///
+/// We follow the approach of: https://iacr.org/archive/pkc2006/39580045/39580045.pdf (Section 5)
```
Comment (akirillo) [inline src/algebra/poly/authenticated_poly.rs:481]:
```suggestion
/// We follow the approach of: https://iacr.org/archive/pkc2006/39580045/39580045.pdf (Section 4)
```

File: src/algebra/poly/authenticated_poly.rs [hunk 2]
```diff
@@ -379,6 +476,65 @@ impl<C: CurveGroup> Div<&DensePolynomialResult<C>> for &AuthenticatedDensePoly<C
...
+/// For f(x) / g(x) where deg(f) = n, deg(g) = m, the objective of a division with
+/// remainder algorithm is to solve:
+///     f(x) = g(x)q(x) + r(x)
+/// for q(x), r(x) uniquely where deg(r) < deg(g).
+///
+/// We could solve for q(x) easily if:
+///     1. We could "mod out" r(x) and
+///     2. If g^{-1}(x) exists
+/// The rev operator provides a transformation that makes both of these true:
+///     rev(f) = rev(g) * rev(q) + x^{n - m + 1} * rev(r)
```
Comment (akirillo) [inline src/algebra/poly/authenticated_poly.rs:495]: the way you defined `rev` at the beginning of this doc comment as "knowing" `deg(a)`, as opposed to the paper's definition `rev_k(a) = x^k * a(1/x)`, makes intuitive sense, but makes the equation above correct only if `deg(r) = m-1`.

we could do the following:
```suggestion
///     rev(f) = rev(g) * rev(q) + x^{n - m + 1} * x^{m-1} * r(1/x)
```

but that doesn't read as nicely, so I'd actually be in favor of my suggestion above on lines 484-485

File: src/algebra/poly/authenticated_poly.rs [hunk 3]
```diff
@@ -379,6 +476,65 @@ impl<C: CurveGroup> Div<&DensePolynomialResult<C>> for &AuthenticatedDensePoly<C
 }
 impl_borrow_variants!(AuthenticatedDensePoly<C>, Div, div, /, DensePolynomialResult<C>, C: CurveGroup);
 
+/// Authenticated division, i.e. division in which the divisor is a secret shared polynomial
+///
+/// We follow the approach of: https://iacr.org/archive/pkc2006/39580045/39580045.pdf (Section 5)
+///
+/// To see why this method holds, consider the `rev` operation for a polynomial a(x):
+///     rev(a) = x^deg(a) * a(1/x)
+/// Note that this operation is equivalent to reversing the coefficients of a(x)
```
Comment (akirillo) [inline src/algebra/poly/authenticated_poly.rs:485]: See my comment below, could we define this as in the paper:
```suggestion
///     rev_k(a) = x^k * a(1/x)
/// Note that this operation is equivalent to reversing the coefficients of a(x) if deg(a) == k
```

File: src/algebra/poly/authenticated_poly.rs [hunk 4]
```diff
@@ -379,6 +476,65 @@ impl<C: CurveGroup> Div<&DensePolynomialResult<C>> for &AuthenticatedDensePoly<C
...
+        let modulus = n - m + 1;
+
+        // Apply the rev transformation
+        let rev_f = self.rev();
+        let rev_g = rhs.rev();
+
+        // Invert `rev_g` in the quotient ring
+        let rev_g_inv = rev_g.mul_inverse_mod_t(modulus);
+
+        // Compute rev_f * rev_g and "mod out" rev_r; what is left is `rev_q`
```
Comment (akirillo) [inline src/algebra/poly/authenticated_poly.rs:529]:
```suggestion
        // Compute rev_f * rev_g_inv and "mod out" rev_r; what is left is `rev_q`
```

**Review by akirillo — APPROVED** [review]
lgtm, mostly just annoying nits and pedantic requests for doc comment changes

File: src/algebra/poly/authenticated_poly.rs (hunk 2 above)
Comment (joeykraut) [reply src/algebra/poly/authenticated_poly.rs:495]: we only ever instantiate it with the degree, no point in adding complexity to an already obscure derivation

**Pull request merged**